File: frontend/src/types/layer.ts
// 지도 렌더링용 (기존 OL 연동)
export type LayerType = 'wmts-base' | 'wms'

export interface LayerItem {
    id: string
    name: string
    type: LayerType
    visible: boolean
    opacity: number
    legend?: string
    wmsLayers?: string
    wmsStyles?: string
}

export interface LayerGroup {
    id: string
    name: string
    children: (LayerItem | LayerGroup)[]
}

export function isLayerGroup(node: LayerItem | LayerGroup): node is LayerGroup {
    return 'children' in node
}

export function flattenItems(nodes: (LayerItem | LayerGroup)[]): LayerItem[] {
    return nodes.flatMap((n) => (isLayerGroup(n) ? flattenItems(n.children) : [n]))
}

// DB 기반 레이어 관리
export type DbLayerType = 'WMS' | 'WMTS' | 'TMS' | 'WFS' | 'MVT' | 'GEOJSON' | 'ARCGIS' | 'XYZ'
export type DbLayerSourceType = 'OPENAPI' | 'GEOSERVER' | 'GEOWEBCACHE' | 'XYZ' | 'STATIC'

export interface DbLayer {
    id: number
    name: string
    type: DbLayerType
    sourceType: DbLayerSourceType
    url: string
    layerName: string | null
    styleName: string | null
    styleConfig: string | null
    format: string | null
    projection: string | null
    minZoom: number | null
    maxZoom: number | null
    opacity: number
    visible: boolean
    sortOrder: number
    groupId: number | null
    groupName: string | null
    description: string | null
    createdAt: string
    updatedAt: string
}

export interface DbLayerFormState {
    name: string
    type: DbLayerType
    sourceType: DbLayerSourceType
    url: string
    layerName: string
    styleName: string
    styleConfig: string
    format: string
    projection: string
    minZoom: string
    maxZoom: string
    opacity: number
    visible: boolean
    sortOrder: number
    groupName: string
    description: string
}

export const LAYER_TYPE_OPTIONS: { value: DbLayerType; label: string }[] = [
    { value: 'WMS',     label: 'WMS' },
    { value: 'WMTS',    label: 'WMTS' },
    { value: 'TMS',     label: 'TMS' },
    { value: 'WFS',     label: 'WFS' },
    { value: 'MVT',     label: 'MVT (벡터 타일)' },
    { value: 'GEOJSON', label: 'GeoJSON' },
    { value: 'ARCGIS',  label: 'ArcGIS REST' },
    { value: 'XYZ',     label: 'XYZ' },
]

export const LAYER_SOURCE_OPTIONS: { value: DbLayerSourceType; label: string }[] = [
    { value: 'OPENAPI',     label: 'OpenAPI' },
    { value: 'GEOSERVER',   label: 'GeoServer' },
    { value: 'GEOWEBCACHE', label: 'GeoWebCache' },
    { value: 'XYZ',         label: 'XYZ' },
    { value: 'STATIC',      label: '정적 파일' },
]

export const EMPTY_LAYER_FORM: DbLayerFormState = {
    name: '', type: 'WMS', sourceType: 'GEOSERVER',
    url: '', layerName: '', styleName: '', styleConfig: '',
    format: 'image/png', projection: 'EPSG:3857',
    minZoom: '', maxZoom: '', opacity: 1, visible: true,
    sortOrder: 0, groupName: '', description: '',
}

// 트리 구조
export interface DbLayerGroup {
    id: number
    name: string
    parentId: number | null
    sortOrder: number
    children: DbLayerGroup[]
    layers: DbLayer[]
}

export interface LayerTreeResponse {
    groups: DbLayerGroup[]
    ungroupedLayers: DbLayer[]
}

// 드래그앤드롭용 flat 아이템
export type TreeNodeType = 'group' | 'layer'

export interface TreeNode {
    id: string           // "group-1" | "layer-1"
    type: TreeNodeType
    depth: number
    parentGroupId: number | null
    sortOrder: number
    groupData?: DbLayerGroup
    layerData?: DbLayer
}

// Permission 옵션
export const PERMISSION_OPTIONS = [
    { value: 'SUPER_ADMIN', label: '슈퍼관리자' },
    { value: 'MAP_ADMIN',   label: '지도관리자' },
    { value: 'VIEWER',      label: 'VIEWER (뷰어)' },
    { value: 'DEPT_A',      label: 'DEPT_A (부서A)' },
    { value: 'DEPT_B',      label: 'DEPT_B (부서B)' },
]

File: backend/src/main/java/com/gtp/domain/map/entity/Layer.java
package com.gtp.domain.map.entity;

import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "tbl_layer")
@EntityListeners(AuditingEntityListener.class)
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Builder
@AllArgsConstructor
public class Layer {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(nullable = false, length = 20)
    private String type; // WMS, WMTS, TMS, WFS, MVT, GEOJSON, ARCGIS, XYZ

    @Column(name = "source_type", nullable = false, length = 20)
    private String sourceType; // OPENAPI, GEOSERVER, GEOWEBCACHE, XYZ, STATIC

    @Column(nullable = false, length = 500)
    private String url;

    @Column(name = "layer_name", length = 200)
    private String layerName;

    @Column(name = "style_name", length = 100)
    private String styleName;

    @Column(name = "style_config", columnDefinition = "TEXT")
    private String styleConfig;

    @Column(length = 50)
    private String format;

    @Column(length = 20)
    private String projection;

    @Column(name = "min_zoom")
    private Integer minZoom;

    @Column(name = "max_zoom")
    private Integer maxZoom;

    @Column(nullable = false)
    @Builder.Default
    private double opacity = 1.0;

    @Column(nullable = false)
    @Builder.Default
    private boolean visible = true;

    @Column(name = "sort_order", nullable = false)
    @Builder.Default
    private int sortOrder = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "group_id")
    private LayerGroup group;

    @Column(name = "group_name", length = 100)
    private String groupName;

    @Column(length = 300)
    private String description;

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public void update(String name, String type, String sourceType, String url, String layerName,
                       String styleName, String styleConfig, String format, String projection,
                       Integer minZoom, Integer maxZoom, double opacity, boolean visible,
                       int sortOrder, LayerGroup group, String description) {
        this.name = name;
        this.type = type;
        this.sourceType = sourceType;
        this.url = url;
        this.layerName = layerName;
        this.styleName = styleName;
        this.styleConfig = styleConfig;
        this.format = format;
        this.projection = projection;
        this.minZoom = minZoom;
        this.maxZoom = maxZoom;
        this.opacity = opacity;
        this.visible = visible;
        this.sortOrder = sortOrder;
        this.group = group;
        this.groupName = group != null ? group.getName() : null;
        this.description = description;
    }

    public void updateSortOrderAndGroup(int sortOrder, LayerGroup group) {
        this.sortOrder = sortOrder;
        this.group = group;
        this.groupName = group != null ? group.getName() : null;
    }
}
